### DISCUSSION TITLE
Default optimization level for Zed extensions

### DISCUSSION THREAD

**Discussion by zamazan4ik** [zed-industries/zed#21450] [Feature Requests]
### Check for existing issues

- [X] Completed

### Describe the feature

Hi!

I know about Zed editor from Reddit posts in the Rust subreddit. What I've learnt their that Zed's devs care **a lot** about the editor's performance. Since I started to use this project recently and I also care about its performance, I want to propose an idea about improving the future performance of the whole Zed ecosystem. I didn't find a proper issue/discussion type for such an idea so feel free to relabel it as you want.

I did a quick overview of multiple popular extensions (the information about them see a bit below) and found that no one enables "advanced" compiler optimizations for them as `codegen-units = 1`, [LTO](https://doc.rust-lang.org/rustc/codegen-options/index.html#lto). I see it as an issue for the Zed ecosystem - the ecosystem that tries to provide the best experience from the performance perspective. I am especially curious about LTO since I am [spending](https://github.com/issues?q=is%3Aissue+author%3Azamazan4ik+archived%3Afalse+LTO+) some time with expanding its usage across Rust. However, my issue is not only about LTO but also about other more aggressive compiler optimization settings for extensions.

Here I collected statistics from some of the most downloaded Zed extensions (according to [this](https://zed.dev/extensions) webpage). All builds are done with x86-64 Fedora 41 + Rust 1.82, with the `cargo build --release` command. The only change between "before" and "after" is enabling `lto = true` for the Release profile:
* https://github.com/rzukic/zed-latex : from 670 Kib to 402 Kib
* https://github.com/zed-extensions/ruby : from 548 Kib to 432 Kib
* https://github.com/zed-extensions/dart : from 516 Kib to 407 Kib
* https://github.com/zed-extensions/vue : from 606 Kib to 416 Kib
* https://github.com/zed-extensions/svelte : from 511 Kib to 434 Kib
* https://github.com/zed-extensions/kotlin : from 501 Kib to 396 Kib
* https://github.com/eth0net/zed-docker-compose : from 460 Kib to 391 Kib
* https://github.com/zed-extensions/swift : from 460 Kib to 400 Kib
* https://github.com/Feel-ix-343/markdown-oxide : from 9.5 Mib to 7.3 Mib

I show binary size improvements because it's much easier to demonstrate compared to performance wins (which requires more work to be done for establishing proper benchmarking scenarios, etc.). According to these tests, LTO consistently improves the binary size of all these extensions.

I decided to use Fat (aka Full) LTO instead of Thin LTO since it allows to perform more aggressive optimizations compared to Thin LTO in the price of longer build times and higher peak memory consumption. However, it shouldn't be a problem for extensions at all for now since they all are pretty small and such an overhead is negligible in practice.

I have several ideas/proposals about how the situation can be improved in this area:
* Integrate into the Zed extensions development [guideline](https://zed.dev/docs/extensions/developing-extensions) a section about performance optimization. As an example - a similar thing from the Tauri [documentation](https://v2.tauri.app/concept/size/#cargo-configuration). It could help developers to apply more aggressive optimizations from a day one for their extensions. However, according to my [findings](https://github.com/issues?q=is%3Aissue+author%3Azamazan4ik+archived%3Afalse+Tauri+) it doesn't completely resolve the issue since people don't read the documentation/they forget to apply the recommended Cargo profile switches to their apps/etc. so this way is not so robust but requires quite little efforts from Zed devs.
*  Implement some kind of (semi-)automatic review of Zed extensions that will check extensions' sources for proper optimization options and warn their devs about missing optimizations. This provides tighter control over extensions' applied optimization but requires more resources for implementation. Maybe some kind of "Extension verification pipeline" could be provided in the future where an optimization check could be one of the steps performed.
* Maybe something else that didn't come yet to my mind.

An option about moving all the extensions into a main repo where you can control the optimization level for all extensions I don't see as a viable option for the community since I am sure that this way is not scalable enough in practice.

One more question to consider if we agree about the importance of such a change in a more "centralized" way - how do we want to apply more compiler optimizations to the existing plugins in the ecosystem? Of course, we can create a zillion issues like "Please enable LTO/codegen-units=1 for your Zed extension" but it won't be too scalable in practice (as you can [see](https://github.com/issues?q=is%3Aissue+author%3Azamazan4ik+archived%3Afalse+LTO+) with my LTO efforts for Rust projects).

Thank you.

P.S. If anyone wonders why Cargo doesn't enable more advanced stuff in its Release profile, you can start reading from [here](https://github.com/rust-lang/cargo/issues/11298).

### Environment

N/A

### If applicable, add mockups / screenshots to help present your vision of the feature

N/A

**Comment by berkus** [comment]
probably goes into the same discussion, but rust-analyzer folks build it without PGO; so to get another 15-20% boost out of rust-only dev performance it would be great if Zed was able to build/bundle better optimized rust-analyzer - see https://github.com/rust-lang/rust-analyzer/issues/9412

**Comment by zamazan4ik** [comment]
Seems like the discussion is forgotten for an unknown reason :) @maxdeviant @notpeter @ConradIrwin @danilo-leal - friendly pinging you as persons who are active in Zed development (just checked GitHub's Insight page). I want to start/continue the discussion about this topic. If you know other interested people in this - please ping them too. Thanks!

**Reply by notpeter** [reply]
Do not ping a bunch of zed staff -- it's notification spam and will likely have the opposite effect that you intend.

**Comment by zamazan4ik** [comment]
Excuse me for such a ping - I just wanted to get any feedback on this proposal (even a rejection). I didn't want to annoy anyone. How can we continue on this in a right way?

**Comment by ConradIrwin** [comment]
@zamazan4ik thanks for the ideas! Adding a recommended settings section to the docs seems reasonable, as does sending PRs to extension authors.

not sure we want to gatekeep extensions on compiler flags, so seems best not to enforce anything by policy/reiview

**Reply by zamazan4ik** [reply]
Thank you for the response!

> Adding a recommended settings section to the docs seems reasonable

If Zed devs agree with this, how can we continue with this? I mean, do I need to create an additional issue/PR/whatever else into the Zed documentation? What is the recommended/desired place to make this information visible and accessible for Zed extensions devs?

> as does sending PRs to extension authors.

Ohh. Unfortunately, this solution is not scalable since I/we need to send a PR into **each** Zed extension if we want to improve the current Zed extensions ecosystem state. I definitely can do it for some extensions, but not for all of them. Is there any existing way to enforce such a thing on Zed level? E.g. like Zed can "automatically" add/inject additional Cargo flags for all extensions on some step before extensions are actually delivered to end users? I don't say that it's an ideal way - just an idea, since I am not familiar enough with a Zed extension lifecycle from source code to a user. On the one hand, tweaking such options in one place is a common thing in similar places like packages in different distributions (like Ubuntu, Fedora and others enforce LTO for packages by default in their package recipes). In this case, we have more control over extensions for the Zed ecosystem → we can deliver more consistent/optimized experience for Zed users (and we care about performance, yeah). On another hand, it requires some additional engineering efforts into this, and I'm pretty sure that for now Zed devs have another priorities.

> not sure we want to gatekeep extensions on compiler flags, so seems best not to enforce anything by policy/reiview

I think it's a good thing to discuss. If the gatekeeping way will be implemented in some automatic way (like the Cargo flags injection way), and it won't disturb extension devs and Zed devs - why not? IMO, we can find such a way ;)

**Reply by notpeter** [reply]
> What is the recommended/desired place to make this information visible and accessible for Zed extensions devs?

We have an example Cargo.toml in our docs here might be a good place to start:
https://github.com/zed-industries/zed/blob/a2e786e0f9c7967083de6a1e82c2d91dcabfc3bd/docs/src/extensions/developing-extensions.md?plain=1#L57-L72

> If the gatekeeping way will be implemented in some automatic way (like the Cargo flags injection way), and it won't disturb extension devs and Zed devs

We currently periodically build all changed extensions from the [zed-industries/extensions](https://github.com/zed-industries/extensions) repo and upload the bundled wasm. So it's certainly possible to mangle everything in a central place, even if it is a little ugly.

**Reply by zamazan4ik** [reply]
> We have an example Cargo.toml in our docs here might be a good place to start

Seems like the right place! I think we can add an additional section where about recommended compiler options, like it's done in Tauri [docs](https://v2.tauri.app/concept/size/#cargo-configuration). I think it will be a good start.

> We currently periodically build all changed extensions from the [zed-industries/extensions](https://github.com/zed-industries/extensions) repo and upload the bundled wasm. So it's certainly possible to mangle everything in a central place, even if it is a little ugly.

Sounds like a good enough way, even if it's a bit ugly! If we agree to go in this way - tweaking Cargo options centrally - what do we need to do next?

**Reply by ConradIrwin** [reply]
PRs welcome!

**Reply by zamazan4ik** [reply]
Where do I need to send a PR to enable building all extensions with LTO by default? I've tried to find this magic build script but found nothing. If the change is pretty straightforward, I am ready to send a PR for that.

**Reply by zamazan4ik** [reply]
@ConradIrwin friendly ping ;)

**Reply by ConradIrwin** [reply]
Not sure. I'd look around at the various `crates/extension*` 

The docs are in `docs/src/extensions` if you wanted to update those too.

**Reply by zamazan4ik** [reply]
Okay, let's start by updating the extensions-oriented documentation and put there a recommendation about enabling LTO and CG1 for plugins.

It would be nice if you will be able to find a script that pulls all changed Zed extensions and rebuild them - I would be happy to integrate LTO and CG1 in such central place too. Right now, I just don't know where this "magic" script is located :(

**Reply by notpeter** [reply]
The tool that builds/packages extensions is `extension_cli` which is [crates/extension_cli/src/main.rs](https://github.com/zed-industries/zed/blob/main/crates/extension_cli/src/main.rs).

**Reply by lnay** [reply]
> PRs to extension authors.

I also think this is best. The extensions that have the most to gain from this with in terms of bundle size and performance will also have the most to lose from increased compile times affecting dev UX (especially in this setting where scripted test suites are not feasible, so you are relying on manual UI testing which cannot happen from the coffee machine).

Ideally you would be able to build the extension with different profiles to mitigate this. I don't know if this is already possible with the cli but this stuff is not documented.

**Reply by zamazan4ik** [reply]
I don't think so, since this solution doesn't scale for all extensions. Manually performing such things is too complicated, developing some (semi)automatic way requires additional development efforts.

That's why I'm suggesting just inject required flags on the Zed side in the `extension` utility level. I quickly investigated the code, and as a quick fix we can add [here](https://github.com/zed-industries/zed/blob/7633bbf55a79d29acd9feb330f17f323a9de800f/crates/extension/src/extension_builder.rs#L160) two additional lines with `CARGO_PROFILE_RELEASE_LTO=true` and `CARGO_PROFILE_RELEASE_CODEGEN_UNITS=1`. If you are okay with such a "fix" - I am ready to create a PR for that.

**Reply by ConradIrwin** [reply]
It seems reasonable. How much slower does it make compiling?

**Reply by zamazan4ik** [reply]
It depends on an extension, but in my experience enabling FatLTO + CG1 doubles the build time of a project in a general case.

**Reply by lnay** [reply]
Admittedly, it did not create multi-minute builds that I was fearing. Clicking "rebuild" for the latex extension in the UI takes about 45 seconds either way for me, whereas cargo builds are much shorter after one line changes (here the dev profile has incremental enabled, and release has all the recommended above):

```shell
$ cargo build --profile=dev --target=wasm32-wasip2
   Compiling latex v0.2.1 ()
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.25s
$ cargo build --profile=release --target=wasm32-wasip2
   Compiling latex v0.2.1 ()
    Finished `release` profile [optimized] target(s) in 1.70s
```

**aside**: It appears that extension building is much more than compiling the wasm?? It could still be nice for the CLI to be more exposed and maybe have options to skip certain things?

**Reply by maxdeviant** [reply]
I'm also of the opinion that we should leave this up to individual extensions to add.

Maybe we can add some linting to warn users who don't have these flags set, but injecting it into the packaging step just in extension CI feels too magical, to me. And it means that extension authors won't see the same things if they were to build their extensions locally.

**Reply by zamazan4ik** [reply]
I agree that the proposed solution is not the best solution from the transparency perspective but I don't see an easier solution here.

> Maybe we can add some linting to warn users who don't have these flags set

This idea has a lot of questions that need to be resolved. How to implement such lints? How to apply these lints to all extensions? How to guarantee that these lints are not ignored by extensions devs? And many more such questions. However, if we apply additional Cargo settings in the `extensions` utility, we can avoid all such problems.

I am not saying that your solution is bad or something like that - just saying that it requires much more development efforts and questions to be resolved.

**Reply by zamazan4ik** [reply]
According to the https://github.com/zed-industries/zed/pull/37524#issuecomment-3254637551 , my tests were not sufficient. And as far as I see this comment:

> It's not even apparent to me whether the changes will result in meaningful wins from a size or performance perspective in a way that matters to users.

Zed team doesn't see enough user value in applying additional optimizations.

@maxdeviant it would be nice if you can describe the exact size bench scenario (with the exact `extension` CLI command line usage) to perform before vs after extension builds to measure size changes with and without LTO. However, as far as I see even with demonstration of such wins you still think that extensions are too small to care about them (and I can understand this point of view).

So, to conclude - right now my understanding is that the Zed dev team is not interested in this idea anymore, and there is no reason to spend/waste time here.

**Reply by lnay** [reply]
For what it's worth @zamazan4ik I think you did a good job raising awareness for this though.

**Comment by notpeter** [comment]
Yo @lnay. It's not ok to spam issues to dozens of repos at once.  It's unprofessional and a violation of our [Code of Conduct](https://zed.dev/code-of-conduct).  

<img width="851" height="1021" alt="Screenshot 2025-09-03 at 15 18 39" src="https://github.com/user-attachments/assets/f2afd38c-13c0-4220-88c3-cccd76bf8a4e" />

The correct path here would have been to propose it to a single extension, with a PR, and then validate the results.  It feels like you didn't get the answer you wanted and so you copy/pasted your complaint 20-odd places.  Don't do this.

**Reply by blopker** [reply]
Ah, that's what happened (https://github.com/blopker/codebook-zed/pull/4). My feeling is that these custom compile options will not have a large enough impact for LSP extensions to justify getting off the defaults. Would love to know if any extension authors learn otherwise though. I just don't know how I would test the performance aspects.

**Reply by lnay** [reply]
Sorry about that, I was imagining individuals only getting single notifications forgetting that a bunch were in the same org. As draft PRs these were mainly intended as suggestions/FYIs.

> My feeling is that these custom compile options will not have a large enough impact for LSP extensions to justify getting off the defaults.

I think the main benefit is bundle size

**Reply by berkus** [reply]
Couldn't there be a way to globally enable this optimization setup without copypasting it across all extensions?

I am sure there was some 21st century technology that allows us to do it more sanely.

**Reply by blopker** [reply]
The defaults are already chosen by expert Rust devs, with a balance of performance, binary size, build time and compatibility.

For sure size is easy to measure, and for larger extensions it may be worthwhile, but the other aspects are harder to test for in the general case.

That being said, LSP extensions are only a handful of KB already. Without a way to measure at least performance impact, which will have a much greater impact on user experience, it's just not worth the risk/effort to change the defaults in the LSP case. Especially since so many codebases will be affected.

I'd love to see some more hard numbers and be wrong though!

**Reply by berkus** [reply]
> The defaults are already chosen by expert Rust devs, 

Could you point to these specific expert devs please?

The defaults do not use good optimization options for production use, only _baseline_.

> but the other aspects are harder to test for in the general case.

Based on the rust knowledge so far, they don't get slower for sure. And every second you save here is mulitplied by whatever number of plugin users are out there, so the savings, however small, completely diminish your expenditure on producing these builds.

> I'd love to see some more hard numbers

Are you implying a classic Catch-22 here?

**Reply by blopker** [reply]
> Could you point to these specific expert devs please?

I guess 'expert' is subjective, but I'm referring to the wonderful team that develops Rust and/or Cargo. I am assuming they have more experience than most.

> Are you implying a classic Catch-22 here?

I was not trying to imply anything, but I can elaborate. My main point is that the defaults are working for many (most?) of these extensions. At ~30k installs, I've personally never had any user complaints of the extension being too large. However, since I'd like to know what effects each option does (also different combinations of options can change outcomes), making these changes does take considerable work in testing.

I'm also not even against doing the work, I just don't even know how I'd measure performance on extensions. Conversely, the bulk of work that Codebook does is in an LSP server which is easily benchmarked, and I have done the work there to find the optimal build configuration for that specific codebase. However, I can say from experience that not every option resulted in better outcomes. 

For example, the PR I just got suggests adding `codegen-units = 1`, which in my previous testing had an immeasurable effect on performance, but increased build time compared to the default of `16`. Maybe it will help with extensions? The point is I don't know, and I don't know how to find out.

If the goal is to have potentially 100's of extension authors make these changes we'd want to be sure to minimize negative outcomes, or at least be aware of them. All I can see at this point is possibly a few of the largest extensions could benefit from LTO binary size reductions, if the increased build time is acceptable for them. Given the small size of Codebook's Zed extension, I can't say it's that compelling to me.

Anyway, if the Zed developers want to bless a particular configuration, I'd be happy to give it a go.

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/community-links
